### ISSUE TITLE
"Token ## not in stream cache" error

### ISSUE THREAD

**Issue by codisms** [dancannon/gorethink#103]
When making multiple async queries to the driver, we sometimes get the above error.  It happens about 20-60% of the calls we're making, so we've had to disable async calls to our API.  The Python driver appears to have had this issue as well and they resolved it:

https://github.com/rethinkdb/rethinkdb/issues/2337

https://github.com/rethinkdb/rethinkdb/commit/ca9ab2835f88a6cf933878eeb3300767e64c9765#diff-d73ba4e9a072dff9a19fe841ca493c8fR228

Is this something that can be address in this project as well?

**Comment by dancannon** [comment]
So I have spent today looking at this and I hope I have fixed the issue. That being said I could not replicate the error using the code in the rethinkdb issue. 

Could you please try the develop branch and let me know what happens?

**Comment by codisms** [comment]
I'll try it, but the commit log doesn't show any changes to that branch.

**Comment by dancannon** [comment]
Ah I should probably push the changes... Try now :)

**Comment by codisms** [comment]
[martini] PANIC: runtime error: slice bounds out of range
/usr/local/Cellar/go/1.3/libexec/src/pkg/runtime/panic.c:552 (0x118fd)
    panicstring: runtime·panic(err);
/usr/local/Cellar/go/1.3/libexec/src/pkg/runtime/panic.c:472 (0x11627)
    panicslice: runtime·panicstring("slice bounds out of range");
/Users/jbailey/Development/go/src/github.com/dancannon/gorethink/results.go:134 (0x32d812)
    (_Cursor).Next: data, c.buffer = c.buffer[0], c.buffer[1:]
/Users/jbailey/Development/go/src/github.com/dancannon/gorethink/results.go:168 (0x32dc08)
    (_Cursor).All: if !c.Next(elemp.Interface()) {

**Comment by dancannon** [comment]
Ok I believe I have fixed this issue (and another race condition). Could you please try the feature/issue_103 branch this time.

If this does not fix your issue would it possible to provide more information about the query you are running.

**Comment by codisms** [comment]
I've uploaded a sample app, and the rethink data we're using, to the repository below.  When I run it with master, I get "gorethink: cannot decode string into Go value of type ..."  With the feature/issue_103 branch, it hangs.

https://github.com/codisms/gorethink-demo

**Comment by dancannon** [comment]
Thanks, I will have a look at that now. Im really sorry that this has been causing you issues. Hopefully this will allow me to fix the bug for good.

**Comment by dancannon** [comment]
Could you please try that branch again, hopefully that should sort it. Your test app works on my machine now.

**Comment by codisms** [comment]
This does appear to be working.

**Comment by dancannon** [comment]
Great thanks for the checking that. I will finish off this and get it merged into master soon.

**Comment by sym3tri** [comment]
+1 also having this issue.

**Comment by dancannon** [comment]
sym3tri, do you still have this issue if you checkout the feature/issue_103 branch? If so could you send in a simple code example or explain what you are doing.

**Comment by sym3tri** [comment]
@dancannon We're basically doing a dump of the entire db. It's happening consistently with a fairly large table > 1million rows. All other smaller tables are fine. 

Can't try that branch yet b/c we're still on v0.2 using an older version of Rethink. Code changes for v0.3 upgrade are in progress.

**Comment by dancannon** [comment]
This fix is now in master, unfortunately due to the significant number of changes required to fix this bug I will not be updating v0.2. If you have any questions about updating to v0.3 let me know.

**Comment by sym3tri** [comment]
@dancannon fyi this did solve the issue for us after upgrading to 1.13.3 + 0.3.1

**Comment by colegleason** [comment]
@dancannon Not exactly sure how this is happening (hard to reproduce consistently), but I am seeing this error again. We are running (as @sym3tri said) 1.13.3 + 0.3.1.

**Comment by dancannon** [comment]
Sorry to hear that you are still having this issue @colegleason, unfortunately I am not sure how I can proceed without more information. I know you are not sure what is causing the bug but could you explain what query you are running and any other information you think might be useful (such as if you are running queries concurrently).

Thanks,
Daniel

**Comment by dancannon** [comment]
Has anybody else had this issue again? I cant seem to replicate it myself.

**Comment by chacken** [comment]
@dancannon after updating my repo today (my previous was from ~1 month ago), I'm getting this all of a sudden.  "gorethink: cannot decode string into Go value of type uint64" Haven't made any changes to my models, etc.. so I'm stumped.  Data inserts fine.. can't pull it back out.

**Comment by dancannon** [comment]
Hi @chacken, I dont suppose you could provide any examples of the data + data structures you are using which cause this error?

**Comment by chacken** [comment]
@dancannon 

Data: &{16986565 chrishacken Chris Hacken 0 false -1 0 3572 283}

Here's my structure.. 100% positive that the ID is a uint64, I created one without an ID though and I still got the error.. so I have no clue what might be going on.  Curious if maybe rethinkdb is storing it as a string.. and then can't decode it back into a uint64?  I also updated rethinkdb to the latest debian package.. so could be something in here or there.

```
type Account struct {
    ID         uint64
    ScreenName string
    Name       string
    EmailCount int64
    InProgress bool
    Cursor     int64
    Progress   int64
    Followers  int64
    Friends    int64
}
func (a Account) FieldMap() map[string]string {
    return map[string]string{
        "ID":         "id",
        "ScreenName": "screen_name",
        "Name":       "name",
        "EmailCount": "email_count",
        "InProgress": "in_progress",
        "Cursor":     "cursor",
        "Progress":   "progress",
        "Followers":  "followers_count",
        "Friends":    "friends_count",
    }
}
```

And here's how I'm trying to select it.

inside some method:

```
var accounts = make([]data.Account,0)
err = storage.SelectAll(&accounts)
```

```
func SelectAll(slicePtr interface{}) error {
  var (
    rows *r.Cursor
    err error
  )

  switch slicePtr.(type) {
  case *[]data.Account:
    rows, err = r.Table("accounts").Run(session)
  }
  if err != nil {
    return err
  }
  err = rows.All(alicePtr)
  if err != nil {
    return err
  }
  return nil
}
```

**Comment by chacken** [comment]
@dancannon Here's a raw example I tried...

http://pastebin.com/raw.php?i=2fP4phqm

data.Account is the same as above.

`chrishacken@nw-twitter:~/go/src/github.com/chacken/nw-twitter$ go run test.go
gorethink: cannot decode string into Go value of type uint64`

P.S. I can create a new issue for this if you want.  Probably not relevant to the above at all.  I searched Google to find this issue.

**Comment by dancannon** [comment]
@chacken I have been really busy the last couple of days and have not been able to look at this issue yet.

I will try to have a look this evening.

**Comment by dancannon** [comment]
@chacken I have had a look into your issue and unfortunately it looks like you are using the `FieldMapper` interface. Unfortunately in my attempt to fix the major performance issues in #125 I removed this feature. 

I recommend using struct tags instead.

**Comment by chacken** [comment]
Don't know how I missed that.  Thanks for your help!  @dancannon
